File: android/app/src/main/java/com/scholarlyapps/pathlingo/ui/utils/GoogleAuthClient.java
package com.scholarlyapps.pathlingo.ui.utils;

import android.content.Context;

import androidx.credentials.CredentialManager;
import androidx.credentials.CredentialManagerCallback;
import androidx.credentials.CustomCredential;
import androidx.credentials.GetCredentialRequest;
import androidx.credentials.GetCredentialResponse;
import androidx.credentials.exceptions.GetCredentialException;

import com.google.android.libraries.identity.googleid.GetSignInWithGoogleOption;
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential;

import java.util.concurrent.Executors;

public class GoogleAuthClient {

    private final Context context;
    private final String webClientId;
    private final CredentialManager credentialManager;

    public GoogleAuthClient(Context context, String webClientId) {
        this.context = context;
        this.webClientId = webClientId;
        this.credentialManager = CredentialManager.create(context);
    }

    public void signIn(GoogleSignInCallback callback) {
        GetSignInWithGoogleOption option = new GetSignInWithGoogleOption.Builder(webClientId).build();
        GetCredentialRequest request = new GetCredentialRequest.Builder()
                .addCredentialOption(option)
                .build();

        credentialManager.getCredentialAsync(
                context,
                request,
                null,
                Executors.newSingleThreadExecutor(),
                new CredentialManagerCallback<GetCredentialResponse, GetCredentialException>() {
                    @Override
                    public void onResult(GetCredentialResponse result) {
                        if (result.getCredential() instanceof CustomCredential) {
                            CustomCredential credential = (CustomCredential) result.getCredential();
                            if (GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL.equals(credential.getType())) {
                                try {
                                    GoogleIdTokenCredential tokenCredential =
                                            GoogleIdTokenCredential.createFrom(credential.getData());
                                    callback.onSuccess(tokenCredential.getIdToken());
                                } catch (Exception ignored) {
                                }
                            }
                        }
                    }

                    @Override
                    public void onError(GetCredentialException e) {
                    }
                }
        );
    }
}

File: android/app/src/main/java/com/scholarlyapps/pathlingo/ui/auth/GoogleAuthClient.kt
package com.scholarlyapps.pathlingo.ui.auth

import android.content.Context
import androidx.credentials.CredentialManager
import androidx.credentials.CustomCredential
import androidx.credentials.GetCredentialRequest
import com.google.android.libraries.identity.googleid.GetSignInWithGoogleOption
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

fun interface GoogleSignInCallback {
    fun onSuccess(idToken: String)
}

class GoogleAuthClient(
    private val context: Context,
    private val webClientId: String,
) {

    private val credentialManager = CredentialManager.create(context)

    fun signIn(callback: GoogleSignInCallback) {
        CoroutineScope(Dispatchers.Main).launch {
            signIn().onSuccess { callback.onSuccess(it) }
        }
    }

    suspend fun signIn(): Result<String> = runCatching {
        require(webClientId.isNotEmpty()) {
            "Missing GOOGLE_WEB_CLIENT_ID. Set it in android/gradle.properties."
        }
        val option = GetSignInWithGoogleOption.Builder(webClientId).build()
        val request = GetCredentialRequest.Builder().addCredentialOption(option).build()

        val response = credentialManager.getCredential(context, request)
        val credential = response.credential
        require(credential is CustomCredential &&
            credential.type == GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL) {
            "Unexpected credential type: ${credential::class.java.name}"
        }
        GoogleIdTokenCredential.createFrom(credential.data).idToken
    }
}
